toothpick-compiler/src/main/java/toothpick/compiler/common/ToothpickProcessor.java

```diff
@@ -234,8 +243,28 @@ protected boolean isValidInjectedType(Element injectedTypeElement) {
             ((TypeElement) enclosingElement).getQualifiedName(), //
             methodOrConstructorElement.getSimpleName(), //
             typeElement);
-        return false;
       }
+      return false;
+    }
+    return true;
+  }
+
+  private boolean isValidProviderOrLazy(Element element) {
+    DeclaredType declaredType = (DeclaredType) element.asType();
+
+    // Contains type parameter
+    if (declaredType.getTypeArguments().isEmpty()) {
+      Element enclosingElement = element.getEnclosingElement();
+      if (enclosingElement instanceof TypeElement) {
+        error(element, "Field %s#%s is not a valid Lazy or Provider.",
```

Comment:
Instead of hardcoding "Lazy or Provider" the type should be printed as in Line 236.